jwst/cube_build/blot_cube_build.py

```diff
@@ -161,145 +166,185 @@ def blot_images(self):
-                weight_distance = 1.0 / weight_distance
-
-                # determine the spaxel xx_cube,yy_cube values of these spaxels
-                # in the ROI so they can be used to pull out the flux of the
-                # median sky cube.
-                yy_cube = (indexr[0] / self.naxis1).astype(np.int)
-                xx_cube = indexr[0] - yy_cube * self.naxis1
-                scf = np.array([self.cube_flux[zz, yy_cube[ir], xx_cube[ir]]
-                                for ir, rr in enumerate(indexr[0]) for zz in indexz[0]])
-                scf = np.reshape(scf, weight_distance.shape)
-                blot_flux[yy, xx] = np.sum(weight_distance * scf)
+                dxy = np.sqrt(dxy)
+                weight_distance = np.exp(-dxy)
+                blot_flux[ydet[ipt], xdet[ipt]] = np.sum(weight_distance * flux_cube[indexr])
                 blot_weight = np.sum(weight_distance)
 
                 # check for blot_weight !=0
                 if blot_weight == 0:
-                    blot_flux[yy, xx] = 0
+                    blot_flux[ydet[ipt], xdet[ipt]] = 0
                 else:
-                    blot_flux[yy, xx] = blot_flux[yy, xx] / blot_weight
-# ________________________________________________________________________________
-            blot.data = blot_flux
-            blot_models.append(blot)
-        t1 = time.time()
-        log.info("Time Blot images = %.1f.s" % (t1 - t0,))
-        return blot_models
+                    blot_flux[ydet[ipt], xdet[ipt]] =  \
+                        (blot_flux[ydet[ipt], xdet[ipt]] / blot_weight)
+
+        #  after all looping
+        # t1 = time.time()
+        # print('Time to call blot_overlap',t1 - t0)
+        return blot_flux
+    # ________________________________________________________________________________
+
+    def blot_overlap_quick(self, xcenter, ycenter, x_cube, y_cube,
+                           flux_cube, blot_flux):
+
+        # blot_overlap_quick finds to overlap between the blotted sky values (x_cube, y_cube)
+        # and the detector pixels. This is faster than blot_overlap.
+        # Looping is done over irregular arrays (x_cube, y_cube) and mapping to xcenter, ycenter
+        # is quicker than looping over detector x,y and finding overlap with large array for
+        # x_cube, y_cube
+        # t0 = time.time()
+        blot_weight = np.zeros_like(blot_flux)
+        num = x_cube.size
+
+        # print('blot_overlap_quick Number of elements on sky mapped to detector going to loop over',num)
+        # TODO need to move this value (roi_det)  to reference file  or at least in spec ?
+        roi_det = 0.5  # 1/2 a pixel
+        for ipt in range(0, num - 1):
+            # search xcenter and ycenter seperately. These arrays are smallsh.
+            # xcenter size = naxis1 (naxis1/2 for MIRI)
+            # ycenter size = naxis2
+            xdistance = np.absolute(x_cube[ipt] - xcenter)
```

Comment:
Are `x_cube` and `y_cube` 1-D arrays or 2-D?  It looks as if they were created by concatenating 1-D arrays together to make 2-D arrays, but `num` is the total number of elements in `x_cube`, not just the number in the first axis.  If they're 1-D arrays, then these differences are scalars, so you don't need `np.where`.  If they're 2-D arrays, it looks to me as if the loop index will eventually exceed the length of the first axis length of `x_cube` and `y_cube`.

Reply:
x_cube and y_cube are 1D arrays. I can not map all the data in one transform but I have to loop over the slices and I just keep adding the inverse transforms from the slice to the x_cube and y_cube.  So xcenter is an array of scalars and x_cube[ipt) is a float it is the exact x detector location which has decimals of the ra,dec, lambda point on the sky.

Reply:
OK; thanks for the clarification.